sdk/core/azure-core/src/md5.cpp

```diff
@@ -0,0 +1,169 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include "azure/core/md5.hpp"
+#include "azure/core/platform.hpp"
+
+#if defined(AZ_PLATFORM_WINDOWS)
+// Windows needs to go before bcrypt
+#include <windows.h>
+
+#include <bcrypt.h>
+#elif defined(AZ_PLATFORM_POSIX)
+#include <openssl/md5.h>
+#endif
+
+#include <stdexcept>
+#include <vector>
+
+namespace Azure { namespace Core {
+
+#if defined(AZ_PLATFORM_WINDOWS)
+
+  namespace Details {
+    struct AlgorithmProviderInstance
+    {
+      BCRYPT_ALG_HANDLE Handle;
+      std::size_t ContextSize;
+      std::size_t HashLength;
+
+      AlgorithmProviderInstance()
+      {
+        NTSTATUS status = BCryptOpenAlgorithmProvider(&Handle, BCRYPT_MD5_ALGORITHM, nullptr, 0);
+        if (!BCRYPT_SUCCESS(status))
+        {
+          throw std::runtime_error("BCryptOpenAlgorithmProvider failed");
+        }
+        DWORD objectLength = 0;
+        DWORD dataLength = 0;
+        status = BCryptGetProperty(
+            Handle,
+            BCRYPT_OBJECT_LENGTH,
+            reinterpret_cast<PBYTE>(&objectLength),
+            sizeof(objectLength),
+            &dataLength,
+            0);
+        if (!BCRYPT_SUCCESS(status))
+        {
+          throw std::runtime_error("BCryptGetProperty failed");
+        }
+        ContextSize = objectLength;
+        DWORD hashLength = 0;
+        status = BCryptGetProperty(
+            Handle,
+            BCRYPT_HASH_LENGTH,
+            reinterpret_cast<PBYTE>(&hashLength),
+            sizeof(hashLength),
+            &dataLength,
+            0);
+        if (!BCRYPT_SUCCESS(status))
+        {
+          throw std::runtime_error("BCryptGetProperty failed");
+        }
+        HashLength = hashLength;
+      }
+
+      ~AlgorithmProviderInstance() { BCryptCloseAlgorithmProvider(Handle, 0); }
+    };
+
+    struct Md5HashContext
+    {
+      std::string buffer;
+      BCRYPT_HASH_HANDLE hashHandle = nullptr;
+      std::size_t hashLength = 0;
+    };
+  } // namespace Details
```

Comment:
I wrote this `AlgorithmProviderInstance` and `Md5HashContext` because storage needs MD5 and SHA-256 and HMAC-SHA256. I just want to abstract all the common logic to avoid duplicate code.

But since you only have MD5 here. It's not necessary any more to have this `AlgorithmProviderInstance` and `Md5HashContext`, so that code can be simplified.

Anyway, it still works, and that's good.